src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/ScopedValue.java

```diff
@@ -0,0 +1,838 @@
+         *
+         * @param key the {@code ScopedValue} key
+         * @param <T> the type of the value
+         * @return the value
+         * @throws NoSuchElementException if the key is not present in this mapping
+         */
+        @SuppressWarnings("unchecked")
+        public <T> T get(ScopedValue<T> key) {
+            var bits = key.bitmask();
+            for (Carrier carrier = this;
+                 carrier != null && containsAll(carrier.bitmask, bits);
+                 carrier = carrier.prev) {
+                if (carrier.getKey() == key) {
+                    Object value = carrier.get();
+                    return (T)value;
+                }
+            }
+            throw new NoSuchElementException();
+        }
+
+        /**
+         * Calls a value-returning operation with each scoped value in this mapping bound
+         * to its value in the current thread.
+         * When the operation completes (normally or with an exception), each scoped value
+         * in the mapping will revert to being unbound, or revert to its previous value
+         * when previously bound, in the current thread.
+         *
+         * <p> Scoped values are intended to be used in a <em>structured manner</em>.
+         * If {@code op} creates a {@link StructuredTaskScope} but does not {@linkplain
+         * StructuredTaskScope#close() close} it, then exiting {@code op} causes the
+         * underlying construct of each {@code StructuredTaskScope} created in the
+         * dynamic scope to be closed. This may require blocking until all child threads
+         * have completed their sub-tasks. The closing is done in the reverse order that
+         * they were created. Once closed, {@link StructureViolationException} is thrown.
+         *
+         * @param op the operation to run
+         * @param <R> the type of the result of the operation
+         * @return the result
+         * @throws Exception if {@code op} completes with an exception
+         * @see ScopedValue#where(ScopedValue, Object, Callable)
+         */
+        public <R> R call(Callable<? extends R> op) throws Exception {
+            Objects.requireNonNull(op);
+            Cache.invalidate(bitmask);
+            var prevSnapshot = scopedValueBindings();
+            var newSnapshot = new Snapshot(this, prevSnapshot);
+            return runWith(newSnapshot, op);
+        }
+
+        /**
+         * Execute the action with a set of ScopedValue bindings.
+         *
+         * The VM recognizes this method as special, so any changes to the
+         * name or signature require corresponding changes in
+         * JVM_FindScopedValueBindings().
+         */
+        @Hidden
+        @ForceInline
+        private <R> R runWith(Snapshot newSnapshot, Callable<R> op) throws Exception {
+            try {
+                JLA.setScopedValueBindings(newSnapshot);
+                JLA.ensureMaterializedForStackWalk(newSnapshot);
```

Comment:
Question: is it necessary here to invoke `ensureMaterializedForStackWalk()` It's really only there to prevent the new `Snapshot` from being scalar replaced. But we know that it cannot be scalar replaced, because it really does escape: a pointer to it is stored in the current `Thread`. So should we simply remove the call to `ensureMaterializedForStackWalk()`, on the grounds that it cannot have any effect?

Reply:
It does seem unnecessary here, but I'm not an expert on current and future C2 escape analysis.  @vnkozlov, do you agree?